### PR TITLE
[Annotations] Add some aria-owns in the text layer to link to annotations (bug 1780375)

### PR DESCRIPTION
**Pull request by calixteman**
This patch doesn't structurally change the text layer: it just adds some aria-owns
attributes to some spans.
The aria-owns attribute expect to have an element id, hence it's why it adds back an
id on the element rendering an annotation, but this id is built in using crypto.randomUUID
to avoid any potential issues with the hash in the url.
The elements in the annotation layer are moved into the DOM in order to have them in the
same "order" as they visually are.
The overall goal is to help screen readers to present to the user the annotations as
they visually are and as they come in the text flow.
It is clearly not perfect, but it should improve readability for some people with visual
disabilities.

### CODE REVIEW

**Comment by calixteman** [comment]
@jcsteh, @MReschenberg, your feedback will be appreciated.

**Comment by Snuffleupagus** [comment]
*Please note:* I'll hold off on reviewing this until feedback has been provided from the a11y-devs, as was requested above, on the off change that the implementation changes significantly.

Furthermore, it'd also be really good to create a new release *first* given the size/scope of these changes. (Which hopefully happens soon, since we agreed to try and have a new release at the end of each month.)

**Comment by timvandermeij** [comment]
Indeed, I'll make a new release tomorrow since it's the end of the month.

**Comment by jcsteh** [comment]
I took a look at this, but I wasn't able to see the annotations with a screen reader. Some questions:

1. Perhaps I'm not understanding how to commit/save annotations to a document. I made a text annotation, dismissed the editor with control+enter, saved the PDF with control+s and then opened the saved PDF in the modified pdf.js. Is there anything else I need to do? I'm pretty sure I can see the annotation with OCR, but it's hard to be certain.
2. Is there anything else I need to do when switching pdf.js branches apart from npm install and gulp server?
3. Are you able to see the annotations in the Firefox Accessibility Inspector?

Also, how will this interact with tagged PDF? I don't fully understand how the text and structure layers fit together, but I know tagged PDF (structure) uses aria-owns as well.

**Comment by jcsteh** [comment]
When the text annotations are added to the DOM, do they have C SS display: none and/or visibility: hidden? That will prevent them from being exposed to a11y.

One other thing I've just noticed is that when a text editor is dismissed, the editor still has role="textbox". That should be removed when contentEditable is disabled. I guess that needs to be handled elsewhere, though?

File: web/pdf_page_view.js
Comment (Snuffleupagus) [inline web/pdf_page_view.js:815]: Most likely this line, or rather its placement, explains the problem mentioned in https://github.com/mozilla/pdf.js/pull/15237#issuecomment-1200655671.

By placing this line here you've limited this functionality to *only* documents with structureTree-information, which I'm guessing isn't actually intended?
If you actually want to call this method *unconditionally*, you'll need to pass in the `AccessibilityManager`-instance to the `TextLayerBuilder`-instance and invoke that method there instead (similar to how the `TextHighlighter`-instance is handled).

Basically, I'm guessing that you want to call it *after* the following line instead: https://github.com/mozilla/pdf.js/blob/803e7af5954b8d84f8dab2b632a3167c7fd14e46/web/text_layer_builder.js#L116

Comment (calixteman) [reply web/pdf_page_view.js]: > I took a look at this, but I wasn't able to see the annotations with a screen reader. Some questions:
> 
>     1. Perhaps I'm not understanding how to commit/save annotations to a document. I made a text annotation, dismissed the editor with control+enter, saved the PDF with control+s and then opened the saved PDF in the modified pdf.js. Is there anything else I need to do? I'm pretty sure I can see the annotation with OCR, but it's hard to be certain.
> 
>     2. Is there anything else I need to do when switching pdf.js branches apart from npm install and gulp server?
>

It's exactly what you've to do.
 
>     3. Are you able to see the annotations in the Firefox Accessibility Inspector?
> 
> 
> Also, how will this interact with tagged PDF? I don't fully understand how the text and structure layers fit together, but I know tagged PDF (structure) uses aria-owns as well.

I fixed the bug mentioned by Jonas and it should be fine now: I can see it correctly in the inspector.
Anyway, for now the content of the annotation is still not available: I just added an `aria-owns` between a span in the text layer and an annotation in the annotation layer. So it should be useful to see at least the annotations which don't have any content like a text field or a checkbox.
I've a wip to make the text content of a FreeText annotation in its HTML counterpart.


Ab

File: src/display/accessibility.js
Comment (Snuffleupagus) [inline src/display/accessibility.js:28]: Should this class perhaps be named `TextAccessibilityManager` instead, and the filename changed to `text_accessibility.js`, to more clearly indicate what it does?

In any case, given how this is now initialized and used, I think this file should be moved into the `web/`-folder instead.

File: src/display/accessibility.js
Comment (Snuffleupagus) [inline src/display/accessibility.js:26]: Nits:
```suggestion
 *  - to reorder elements in the DOM with respect to the visual order;
 *  - to create a link, using aria-owns, between spans in the textLayer and
 *    annotations in the annotationLayer. The goal is to help to know
 *    where the annotations are in the text flow.
```

File: src/display/accessibility.js
Comment (Snuffleupagus) [inline src/display/accessibility.js:175]: Having to manually edit the code, in order to enable debugging, seems less than ideal.

The elements in question are always placed in the textLayer, right?
Because if so, it'd be *much better* to piggyback on the existing `textLayer=visible` debugging hash-parameter with appropriate rules added in the `web/debugger.css` file.

File: src/display/display_utils.js
Comment (Snuffleupagus) [inline src/display/display_utils.js:29]: What's the performance of this code like?

---

Given that we use a hard-coded prefix string with the Editing-functionality, I suppose that we *could try* a similar approach here as well if that'd make things simpler and more efficient. (As long as the resulting `id` is *unlikely* to clash with named destinations.)

File: web/pdf_page_view.js
Comment (Snuffleupagus) [inline web/pdf_page_view.js:188]: We probably don't want to initialize this unconditionally, since the textLayer may not actually be used (depending on options). It ought to suffice, as far as I can tell, to initialize it a the top this block: https://github.com/mozilla/pdf.js/blob/4f6cd05a53e6c069bb4ebee469fda6437a0971af/web/pdf_page_view.js#L699

File: web/text_layer_builder.js
Comment (Snuffleupagus) [inline web/text_layer_builder.js:122]: Generally speaking, let's make the `AccessibilityManager` a bit more similar to the `TextHighlighter` class.
Hence, let's call this method `enable` instead. (You might also want to introduce a `disable` method, to handle any clean-up when the textLayer goes away.)

Comment (Snuffleupagus) [reply web/text_layer_builder.js:122]: Another thing, that could perhaps simplify things and reduce the number of DOM lookups in the `AccessibilityManager`-code, would be to again follow the `TextHighlighter`-class and pass in the textLayer-divs similar to: https://github.com/mozilla/pdf.js/blob/4f6cd05a53e6c069bb4ebee469fda6437a0971af/web/text_layer_builder.js#L99

File: src/display/accessibility.js
Comment (Snuffleupagus) [inline src/display/accessibility.js:40]: Nit: This can be removed.

**Review by Snuffleupagus — COMMENTED** [review]
*Leaving a few basic comments about the overall shape/structure of the new `AccessibilityManager`-class, since I don't expect the a11y-review to really change that.*

File: src/display/accessibility.js
Comment (calixteman) [reply src/display/accessibility.js:175]: Excellent idea !

File: src/display/display_utils.js
Comment (calixteman) [reply src/display/display_utils.js:29]: I ran this on my powerful computer:
```
s = performance.now(); for (let i = 0; i < 1e7; i++) { crypto.randomUUID(); }; d = performance.now() - s 
```

and I got `9475ms` so a unique call should be around `1ms`, so in considering that it's called one time I think it's acceptable.
That said, it's on my powerful computer...

I use the ids in some tests (here and in m-c (i.e. https://searchfox.org/mozilla-central/search?q=%23pdfjs_internal_editor_&path=&case=false&regexp=false)). 
I think we could use `isInAutomation` and/or preprocessing stuff we have here to generate a specific id for test or add an attribute... I'm not sure that's a good idea...

File: web/pdf_page_view.js
Comment (jcsteh) [reply web/pdf_page_view.js]: > Anyway, for now the content of the annotation is still not available: I just added an aria-owns between a span in the text layer and an annotation in the annotation layer. So it should be useful to see at least the annotations which don't have any content like a text field or a checkbox.

I'm not really sure how to test this/provide feedback if text annotations aren't fully exposed yet, as I can't create other kinds of annotations.

Comment (calixteman) [reply web/pdf_page_view.js]: Once this PR is merged (it should be in 30 minutes):
https://github.com/mozilla/pdf.js/pull/15267

it should be a way easier to test this PR.

Comment (calixteman) [reply web/pdf_page_view.js]: @jcsteh, I merged the other PR and I rebased this one, so we should be good now.
Sorry for the inconvenience: I should have written the second patch in first.

**Comment by jcsteh** [comment]
Thanks. I tested this with both a tagged and an untagged PDF and I can now see the text annotations. Nice!

Ideally, I think the annotations would have role="insertion" or similar so you can see they're annotations (as opposed to part of the normal text). It's tricky to know what role is best here; some might argue these are more like comments than suggested insertions, in which case role="comment" is probably better. Regardless, I don't think this should block this; it's a nice-to-have and could be done in a follow-up.

File: web/text_accessibility.js
Comment (Snuffleupagus) [inline web/text_accessibility.js:51]: Similar to the `TextHighlighter`-class, why can't you add a `setTextMapping`-method in this class instead and call that one using something like the following
```javascript
this.accessibilityManager?.setTextMapping(this.textDivs);
```
placed just *after* [this line](https://github.com/mozilla/pdf.js/blob/e1e97c6edd8729e968bde47f60a1901e8e649901/web/text_layer_builder.js#L99)?

That way, the DOM elements should be directly available without the need for any additional DOM lookups here.

File: src/display/editor/tools.js
Comment (calixteman) [reply src/display/editor/tools.js:56]: It was to be sure to have an unique id and make it findable (with a query selector) in tests.

File: web/text_accessibility.js
Comment (calixteman) [reply web/text_accessibility.js:45]: Some elements could have been added before the text layer finished to render, and when `enable` is called those elements are really inserted.
So we need to know that the text layer is still rendering in order to push those elements in a queue.

Comment (Snuffleupagus) [reply web/text_accessibility.js:45]: > So we need to know that the text layer is still rendering in order to push those elements in a queue.

But the `enable`-call occurs at the point when the textLayer has finished rendering, so you already know when that happens. Perhaps what you really want here is an `#enabled`-property, similar to the `TextHighlighter`-class, that you check instead?

File: web/pdf_page_view.js
Comment (Snuffleupagus) [inline web/pdf_page_view.js:701]: Given how the textLayer is being re-created on zooming/rotation, but the annotationLayer and annotationEditorLayer are kept, will this work correctly as-is?
With this code the annotationLayer/annotationEditorLayer will potentially contain a reference to an "old" `TextAccessibilityManager`, which seems like it probably doesn't work.

Can we just re-use the same `TextAccessibilityManager`-instance, but still lazily initialized, with the following code?
```suggestion
      this._accessibilityManager ||= new TextAccessibilityManager(div);
```

Comment (calixteman) [reply web/pdf_page_view.js:701]: I found this bug.

File: src/display/annotation_layer.js
Comment (Snuffleupagus) [inline src/display/annotation_layer.js:2480]: Can you please elaborate what, and why, all these z-index changes are for?

File: src/display/annotation_layer.js
Comment (Snuffleupagus) [inline src/display/annotation_layer.js:2524]: I wonder we perhaps should add a (private) helper method for this code, to reduce the risk of future bugs, rather than effectively repeating this twice?

File: src/display/editor/annotation_editor_layer.js
Comment (Snuffleupagus) [inline src/display/editor/annotation_editor_layer.js:524]: Given that the `accessibilityManager` is also used for the "regular" annotationLayer, which may not be removed at the same time as this editor, is it actually correct to have this `disable`-call here?

File: web/text_accessibility.js
Comment (Snuffleupagus) [inline web/text_accessibility.js:83]: Do you perhaps want code similar to the `TextHighlighter` here as well?

https://github.com/mozilla/pdf.js/blob/ab1297f0538c51e8e7ece037768e38a9991dcc37/web/text_highlighter.js#L64-L70

File: web/text_accessibility.js
Comment (Snuffleupagus) [inline web/text_accessibility.js:108]: Do you perhaps want code similar to the TextHighlighter here as well?
https://github.com/mozilla/pdf.js/blob/ab1297f0538c51e8e7ece037768e38a9991dcc37/web/text_highlighter.js#L86-L89

File: web/text_accessibility.js
Comment (Snuffleupagus) [inline web/text_accessibility.js:84]: Why do we need the timeout here, maybe add a comment if it's indeed necessary?

File: src/display/annotation_layer.js
Comment (calixteman) [reply src/display/annotation_layer.js:2480]: Without this patch annotations are rendered in the order they're in the pdf.
With this patch they're reordered when they're inserted in order to respect the visual order, hence an annotation which should be above an other one, could now be underneath.
The z-index will help to respect the order of the annotation rendering.

Comment (Snuffleupagus) [reply src/display/annotation_layer.js:2480]: Thanks for the explanation; some sort of (short) comment about that might not hurt :-)

Comment (calixteman) [reply src/display/annotation_layer.js:2480]: Yep I added one.

File: src/display/annotation_layer.js
Comment (Snuffleupagus) [inline src/display/annotation_layer.js:1905]: Couldn't such a large value cause a Popup to "interfere" with UI elements such as the findbar and secondaryToolbar (given their z-index definitions)?
You might want to limit this value, and the one below, to only `1000` instead since the likelihood of any one page having (even close to) that many annotation should be extremely low.

File: src/display/annotation_layer.js
Comment (Snuffleupagus) [inline src/display/annotation_layer.js:2469]: Given that the most important thing here is adding the annotation-element to the DOM, how about the following name instead to make that clearer (and if possible, let's make this method private)?
```suggestion
  static #appendElement(element, id, div, accessibilityManager) {
```

**Review by Snuffleupagus — APPROVED** [review]
r=me, with one more (small) comment and all tests passing; thank you!

File: web/annotation_layer_builder.css
Comment (Snuffleupagus) [inline web/annotation_layer_builder.css:213]: *Another thing that just occurred to me, although I've not tested it:*
Do you actually need this property, now that an explicit z-index is assigned in the JS code?

Comment (calixteman) [reply web/annotation_layer_builder.css]: It seems useless: I removed it.
I changed:
https://github.com/mozilla/pdf.js/pull/15237/files#diff-d7b040092cf6be09a276b785ba03c46e79bc8342cf4ae4fb1bc139dea8513425R1906

because the `+=` was just concatenating "1000" to the zindex string when the `-=` below was subtracting 1000.

**Comment by calixteman** [comment]
The "regressions" are a matter of z-index when several popups are displayed.
As said in an other PR, I think we should have a strategy to position a popup depending on its position on the page and maybe on the positions of other open popups.
Anyway, I think those "regressions" are ok, @Snuffleupagus is it ok for you either ?

**Comment by calixteman** [comment]
Yep I tested the affected documents and it was ok.
I had an other look just to be sure that nothing was wrong and I found something in `issue14438.pdf`:
image
the section wrapping the popup has some "large" dimensions, hence the other popups cannot catch the mouse events:
image
And I noticed they're few empty other sections:
image
which should likely removed or at least made invisible because they contribute to increase the viewport width and consequently a vertical scrollbar can appear.

I think the change around z-index just shows a pre-existing issue: an other ordering in the pdf should have shown the same issue.

File: web/text_accessibility.js
Comment (calixteman) [reply web/text_accessibility.js:92]: Good catch.

**Comment by calixteman** [comment]
> Yep I tested the affected documents and it was ok. I had an other look just to be sure that nothing was wrong and I found something in `issue14438.pdf`: image the section wrapping the popup has some "large" dimensions, hence the other popups cannot catch the mouse events: image And I noticed they're few empty other sections: image which should likely removed or at least made invisible because they contribute to increase the viewport width and consequently a vertical scrollbar can appear.
> 
> I think the change around z-index just shows a pre-existing issue: an other ordering in the pdf should have shown the same issue.

A way to fix those issues, would be to set the dimension to `auto` for those specific annotation (popupAnnotation). Wdyt ?

**Comment by Snuffleupagus** [comment]
> A way to fix those issues, would be to set the dimension to `auto` for those specific annotation (popupAnnotation). Wdyt ?

I'd be a bit surprised if there's not other fallout from such a change, so probably not somewhat that we want to attempt here.

---

Let's land this as-is, since hopefully most issues should've been found now :-)

**Pull request merged**
